crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,813 @@
+    }
+}
+
+fn _zxz_decomp_svd(a: DMatrix<Complex64>) -> (DMatrix<Complex64>, DMatrix<Complex64>) {
+    let (v, sigma, w_dg) = svd_decomposition(&a);
+
+    let s = &v * &sigma * &v.adjoint();
+    let u = v * w_dg;
+
+    (s, u)
+}
+
+/// Block ZXZ decomposition method, by Krol and Al-Ars [2]
+fn _block_zxz_decomp(
+    mat: &DMatrix<Complex64>,
+) -> (
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+    DMatrix<Complex64>,
+) {
+    debug_assert!(verify_unitary(mat));
+
+    let i = Complex64::new(0.0, 1.0);
+    let n = mat.shape().0 / 2;
+    let x = mat.view((0, 0), (n, n));
+    let y = mat.view((0, n), (n, n));
+    let u21 = mat.view((n, 0), (n, n));
+    let u22 = mat.view((n, n), (n, n));
+    let (sx, ux) = _zxz_decomp_svd(x.into_owned());
+    let (sy, uy) = _zxz_decomp_svd(y.into_owned());
+    let c = ((&uy.adjoint() * &ux) * i).adjoint();
+    let a1 = (sx + sy * i) * &ux;
+    let a2 = u21 + (u22 * (uy.adjoint() * ux) * i);
+    let b = (a1.adjoint() * x) * Complex64::from(2.0) - DMatrix::<Complex64>::identity(n, n);
+    (a1, a2, b, c)
+}
+
+/// Verify ZXZ decomposition gives the same unitary
+#[allow(clippy::toplevel_ref_arg)]
+fn _zxz_decomp_verify(
+    mat: &DMatrix<Complex64>,
+    a1: &DMatrix<Complex64>,
+    a2: &DMatrix<Complex64>,
+    b: &DMatrix<Complex64>,
+    c: &DMatrix<Complex64>,
+) -> bool {
+    let n = mat.shape().0 / 2;
+    let iden = DMatrix::<Complex64>::identity(n, n);
+
+    let a_block = stack![a1, 0; 0, a2];
+
+    let b1 = iden.clone() + b;
+    let b2 = iden.clone() - b;
+    let b_block = stack![b1, b2; b2, b1];
+
+    let c_block = stack![iden, 0; 0, c];
+
+    let mat_check = a_block * b_block * c_block * Complex64::from(0.5);
+
+    const EPS: f64 = 1e-7;
+    abs_diff_eq!(mat, &mat_check, epsilon = EPS)
+}
+
+///  Decompose a generic multiplexer.
+///
+///        ────□────
+///         ┌──┴──┐
+///       /─┤     ├─
+///         └─────┘
+///
+/// represented by the block diagonal matrix
+///
+///         ┏         ┓
+///         ┃ um0     ┃
+///         ┃     um1 ┃
+///         ┗         ┛
+///
+/// to
+///            ┌───┐
+///     ───────┤ Rz├──────
+///       ┌───┐└─┬─┘┌───┐
+///     /─┤ w ├──□──┤ v ├─
+///       └───┘     └───┘
+///
+/// where v and w are general unitaries determined from decomposition.
+///
+/// .. note::
+///
+///     When the original unitary is controlled then the default value of ``opt_a2 = False``
+///     as we start with the demultiplexing step that does not work with the optimization A.2 of [1, 2].
+fn demultiplex(
+    um0: &DMatrix<Complex64>,
+    um1: &DMatrix<Complex64>,
+    opt_a1: bool,
+    opt_a2: bool,
+    depth: usize,
+    _ctrl_index: Option<usize>,
+    vw_type: VWType,
+) -> PyResult<(CircuitData, DMatrix<Complex64>, DMatrix<Complex64>)> {
+    let um0 = closest_unitary(um0.clone());
+    let um1 = closest_unitary(um1.clone());
+
+    debug_assert!(verify_unitary(&um0));
+    debug_assert!(verify_unitary(&um1));
```

Comment:
if we have to call `closest_unitary` here, maybe we don't need `debug_assert!` ?